### PR TITLE
Database objects periodic import

### PR DESCRIPTION
**Pull request by Tener**
This PR adds periodic import of database objects from supported databases.

Currently implemented for Postgres, which necessitated some refactoring.

Updates https://github.com/gravitational/teleport/issues/32627

Changelog: Automatically import database objects from supported databases per configured import rules.

### CODE REVIEW

File: lib/srv/db/postgres/objects.go
Comment (greedy52) [inline lib/srv/db/postgres/objects.go:42]: hmm. `default_database` is not really required for Postgres auto-user provisioning, and not used at all for the current DAC implementation IIRC.

Maybe pick a database name from the non-wildcard rules? Or default to "postgres" database?

File: lib/srv/db/postgres/connector.go
Comment (greedy52) [inline lib/srv/db/postgres/connector.go:58]: why not copy like this:
https://github.com/gravitational/teleport/blob/0f4eadd8d0a34eb6119e58bc2d9cb3bfc3e87b4b/lib/srv/db/common/session.go#L87-L88
Are we explicitly skipping `log`?

File: lib/srv/db/objects/importer.go
Comment (Tener) [reply lib/srv/db/objects/importer.go:75]: > currently imp.start is called twice so maybe change startSingleDatabaseImporter to a constructor.
Oops, that should not be the case. I'll fix. I actually had `newSingleDatabaseImporter`, which is why that call is there - a leftover - but it made passing the cancel context somewhat fragile, which is why I went with this solution.

File: lib/srv/db/objects/importer.go
Comment (Tener) [reply lib/srv/db/objects/importer.go:95]: No, as there may be other importers running, the service may only be down temporarily etc. In general I believe it is better to keep these objects around rather than delete them unnecessarily.

File: lib/srv/db/objects/importer.go
Comment (Tener) [reply lib/srv/db/objects/importer.go:54]: That is part of it, sure. I'll add a comment.

File: lib/srv/db/server.go
Comment (Tener) [reply lib/srv/db/server.go:87]: This is done on demand in the unit test: `TestStartDatabaseImporter`

File: lib/srv/db/objects/importer.go
Comment (Tener) [reply lib/srv/db/objects/importer.go:202]: The expiry check is part of the reason: we could pass custom comparer that ignores expiry and it wouldn't send any updates if only the expiry is unchanged.

But we would need to track the expiry in some way anyway and I didn't see a satisfactory way forward.

Overall, I think `services.Reconciler` isn't quite suited for this update strategy and implementing the core logic from scratch leads to better readability. We might make the code generic if in the future if there will be more suitable use cases for it.

File: lib/srv/db/objects/objects.go
Comment (Tener) [reply lib/srv/db/objects/objects.go:86]: Agreed, I'll add.

File: lib/srv/db/postgres/objects.go
Comment (Tener) [reply lib/srv/db/postgres/objects.go:42]: Hmm, I'm not certain how safe that heuristic is, but I like the idea of reducing the configuration requirements. What about trying either `postgres` or `teleport-admin`?

File: lib/srv/db/postgres/connector.go
Comment (Tener) [reply lib/srv/db/postgres/connector.go:58]: Ah, yes that is much better. Harder to forget about fields in this way 😅

File: lib/srv/db/objects/fetcher.go
Comment (Tener) [reply lib/srv/db/objects/fetcher.go:75]: Hmm, in general I'm not a fan of tying the control flow to the error values, but I'll see if this can be improved.

File: lib/srv/db/postgres/objects.go
Comment (greedy52) [reply lib/srv/db/postgres/objects.go:42]: Let's do `postgres` so users don't have to create a new database. we can make the doc clear that either the user gives admin user permission to connect `postgres` or specify a default database.

File: lib/srv/db/objects/importer.go
Comment (greedy52) [reply lib/srv/db/objects/importer.go:95]: sounds good. Let's add a comment on that. i guess that could be part of the future performance improvement.

That said, should `s.cfg.DatabaseObjects.StopImporter` be called in `s.stopProxyingDatabase`? For example, if the `server.Shudown` is called, the importer is not stopped with current impl. IIRC `server.Shudown` can be called during a CA rotation or a systemd reload too.

File: lib/srv/db/objects/fetcher.go
Comment (greedy52) [reply lib/srv/db/objects/fetcher.go:75]: > Hmm, in general I'm not a fan of tying the control flow to the error values, but I'll see if this can be improved.

me neither. but I prefer that over returning `nil, nil`. 

Or instead of `trace.NotImplemented`, do a precheck before calling this and return a hard `trace.NotFound` and don't handle it? Or maybe add a bool in this return or something.

File: lib/srv/db/postgres/objects.go
Comment (Tener) [reply lib/srv/db/postgres/objects.go:42]: Makes sense, I'll do so.

File: lib/srv/db/objects/importer.go
Comment (Tener) [reply lib/srv/db/objects/importer.go:95]: Wouldn't the server shutdown cancel the parent context? I'll check out `stopProxyingDatabase` too.

File: lib/srv/db/objects/fetcher.go
Comment (Tener) [reply lib/srv/db/objects/fetcher.go:75]: We can check if given protocol is implemented, sure, but the second case for nil fetcher is a particular database configuration that may not support fetching objects.  These are very similar in the end.

File: lib/srv/db/objects/importer.go
Comment (Tener) [reply lib/srv/db/objects/importer.go:246]: Looking at this with fresh eyes I think some refactoring is due. This code piece is lifted from Postgres DAC implementation, but it would be good to abstract away this pattern (i.e. fetch rules, fetch objects, apply rules, return finished objects).

File: lib/srv/db/objects/importer.go
Comment (greedy52) [reply lib/srv/db/objects/importer.go:95]: > Wouldn't the server shutdown cancel the parent context? I'll check out `stopProxyingDatabase` too.

good point. as long as it gets stopped during ShutDown, we are good.

File: lib/srv/db/objects/fetcher.go
Comment (Tener) [reply lib/srv/db/objects/fetcher.go]: I've simplified this a lot; basically, we are down to standard error handling with the custom bit being lowering the severity of log in case the database is of unsupported protocol.

File: lib/srv/db/objects/importer.go
Comment (Tener) [reply lib/srv/db/objects/importer.go]: Refactored to `go imp.start(ctx)` as well as bunch of other related changes.

File: lib/srv/db/objects/importer.go
Comment (Tener) [reply lib/srv/db/objects/importer.go]: I've done the refactoring I've mentioned above. All object-fetching code is using the same interface now, with import rules being applied consistently by a wrapping fetcher.

File: lib/srv/db/postgres/objects.go
Comment (Tener) [reply lib/srv/db/postgres/objects.go]: Done: the requirement for db name was dropped, we default to "postgres" in "get all db names" function.

File: lib/srv/db/objects/objects.go
Comment (Tener) [reply lib/srv/db/objects/objects.go:127]: Added, including the option to disable the service entirely.

File: lib/srv/db/postgres/users.go
Comment (greedy52) [inline lib/srv/db/postgres/users.go:201]: I really like reusing the same fetcher for user sessions. Is it possible to reuse the same admin connection for calling the permission procedures though?

Comment (Tener) [reply lib/srv/db/postgres/users.go:201]: Reuse sounds tricky to do properly. I'll look into it, I do have an idea that might work. Even if not, perhaps this is acceptable for the ease of maintenance this otherwise brings.

Comment (greedy52) [reply lib/srv/db/postgres/users.go:206]: could we cache it in the `fetcher`? something like

```
type cachedAdminPGXConnGetter interface {
  getCachedPGXAdminConn() (*pgx.Conn, error)
}

if cached, ok := fetcher.(cachedAdminPGXConnGetterr); ok {
  conn, err = cached.getCachedAdminConn()
} else {
  conn, err = e.connectAsAdminSessionDatabase(ctx, sessionCtx)
} 
```
Or maybe move the apply part to the fetcher interface as well. I will leave it up to you if you want to keep it like this or we can tackle this later when doing performance improvement.

File: lib/srv/db/server.go
Comment (greedy52) [inline lib/srv/db/server.go:491]: Are we still getting this when postgres databases are not configured with auto-user provisioning?

**Review by greedy52 — APPROVED** [review]

File: lib/srv/db/server.go
Comment (Tener) [reply lib/srv/db/server.go:491]: No, implementation-wise `NotImplemented` is only returned for unsupported protocols.

File: lib/srv/db/postgres/users.go
Comment (Tener) [reply lib/srv/db/postgres/users.go:206]: Yep, I've tried something along these lines. Trouble is with ensuring the connection params are the same, which often isn't the case. I'll give it another look tomorrow though, perhaps caching the fetcher is indeed the way forward.

File: lib/srv/db/objects/objects.go
Comment (Tener) [reply lib/srv/db/objects/objects.go]: Done.

**Comment by Tener** [comment]
@timothyb89 do you mind giving it a look?

File: lib/srv/db/postgres/users.go
Comment (greedy52) [reply lib/srv/db/postgres/users.go:206]: >  I'll give it another look tomorrow though, perhaps caching the fetcher is indeed the way forward.

I just gave it another thought that it won't work well as we have to deal with closing the connection.

File: lib/srv/db/server.go
Comment (greedy52) [reply lib/srv/db/server.go:491]: > No, implementation-wise `NotImplemented` is only returned for unsupported protocols.

a postgres database without admin user will return BadParameter and come here?

File: lib/srv/db/postgres/users.go
Comment (Tener) [reply lib/srv/db/postgres/users.go:206]: Yes, that is another complication I've run into. I've considered using a connection pool instead, but those are not designed for our use case at all.

File: lib/srv/db/server.go
Comment (Tener) [reply lib/srv/db/server.go:491]: Yes, this will give a warning. Do you think we should handle that case separately?

Comment (greedy52) [reply lib/srv/db/server.go:491]: i mean not setting admin user is a common case, not an error case. we shouldn't error/warn in that case, or maybe we shouldn't even start importer when admin user is not set.

Comment (Tener) [reply lib/srv/db/server.go:491]: I took it from the perspective of answering the question of "why does the importer doesn't work for my database", but you are right, this doesn't deserve a treatment of an error. I'll add handling for this.

Comment (Tener) [reply lib/srv/db/server.go:493]: We will now get this sort of message (debug verbosity):
```
2024-07-12T15:20:30+02:00 DEBU [DB:SERVIC] Database object importer cannot be started due to disabled fetcher reason:no admin user configured database:mypg db/server.go:491
```

**Review by timothyb89 — APPROVED** [review]

File: lib/srv/db/objects/objects.go
Comment (GavinFrazar) [inline lib/srv/db/objects/objects.go:87]: could you add a comment explaining why the numbers 12 and 3?
It may be a good idea also to add the unit of time to these env vars to make it clear, e.g. `TELEPORT_UNSTABLE_DB_OBJECTS_OBJECT_TTL_MINUTES` (I'm actually not sure what the units are assumed to be currently)

edit: ok I see `time.ParseDuration` handles a wide range of time formats, including negative values.
I think I still prefer something like `_MINUTES` and just parsing an unsigned int, with a check for the 0 value if 0 is invalid. That enforces a minimum value of 1 minute automatically, and is simpler to reason about plus the env var name documents itself.

**Review by GavinFrazar — APPROVED** [review]

File: lib/srv/db/objects/objects.go
Comment (Tener) [reply lib/srv/db/objects/objects.go:135]: I'm not sure about minimum values. In general, the env var overrides are there for very low-level access. Good point about avoiding panics, however. I've modified the `disabled()` method to check for zero or negative interval.

File: lib/srv/db/objects/objects.go
Comment (Tener) [reply lib/srv/db/objects/objects.go:89]: I think I picked this pattern from elsewhere in Teleport, feels good to keep to one convention. The updated logic checks for zero/negative intervals which matter, so we should be good. In any case messing with these `UNSTABLE` env vars puts responsibility on the administrator to pick sensible values and monitor the stability afterwards.

**Comment by Tener** [comment]
@r0mant this PR needs admin approval, can you provide one?

**Review by r0mant — APPROVED** [review]
Bot.

**Pull request merged**
